dom/src/main/scala/fs2/dom/Event.scala

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2022 Arman Bilge
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fs2.dom
+
+import cats.effect.kernel.Sync
+import org.scalajs.dom
+
+import scala.concurrent.duration._
+
+abstract class Event[F[_]] private[dom] {
```

Comment:
Why not using opaque types here for the same reason as for the rest of the dom api?

Reply:
It's a good question.

The reason is, I generally consider it to be an anti-pattern when you need a constraint to use methods on the instance of an object e.g. all the methods on `HtmlElement` require the `Dom` constraint. Contrast with `Ref`, `Deferred`, `Queue`, `Signal`, http4s `Client`, etc. where once you have an instance, you can always use the methods.

The only reason I made an exception for `HtmlElement` is because a typical application may need to make very many of them very quickly, and the overhead of a wrapper would cause a lot of allocations.

However, `Event`s fire more rarely, so I think a wrapper is an acceptable cost, and being able to use the methods without an extra constraint is better ergonomics IMO.

Actually, I even am entertaining ideas to solve this problem for `HtmlElement` too. Even though making a wrapper class is too expensive, because this is JS we can stuff an `Async` constraint into a dynamic property. But I hesitate, because it's a bit too hacked 😅 

But, I am curious to hear your opinion on this issue :)

Reply:
> The reason is, I generally consider it to be an anti-pattern when you need a constraint to use methods on the instance of an object

:+1:, at least when they all require the same constraint.

> However, Events fire more rarely, so I think a wrapper is an acceptable cost, and being able to use the methods without an extra constraint is better ergonomics IMO

Ok, I can see the general reasoning now. I think in most cases this is right. However in more edge cases like games there can be lots of events. But maybe there is an even smaller bottleneck for these cases, who knows. The main problem with this is it's hard to decide when it really pays to stress for minimizing allocations. It's way easier when you _know_ you're too slow :)

> Actually, I even am entertaining ideas to solve this problem for HtmlElement too

Don't think it's worth it at current project phases. Users will hardly recognize. FT people are happy to pass `Dom` constraint down the tree and IO people get it "for free". I would prioritize feature completeness and bugs atm. SVG and "reading foreign html" are the hottest for me. It leaves a bad taste for newcomers when they can't express themselves :)

Reply:
> However in more edge cases like games there can be lots of events.

True. But in those cases, do you _really_ want to be using `IO`? At the end of the day, we are talking one allocation to wrap the event, compared to multiple `IO` allocations to then _do_ something with the event :)

Also, probably you are only processing a few events at a time, but `HtmlElement`s can be very long-lived and if they are wrapped make extra memory pressure for the entire duration of the application.

--- 

> It leaves a bad taste for newcomers when they can't express themselves :)

What's this 😅 

> SVG and "reading foreign html" are the hottest for me

Definitely open some issues about those :) esp about the foreign HTML.

Reply:
> What's this

I mean for example if a user wants his icons to be drawn using SVG but he can't, doesn't make him happy ;)

Reply:
> But in those cases, do you really want to be using IO?

Of course, except performance is really unacceptable :smile: